### PR TITLE
Compute workspace ancestor / descendant / ordered dependency graph

### PR DESCRIPTION
**Pull request by cristiano-belloni**
# Dependency resolver API

## computeAncestorSet

From one or more vertices in a graph, compute the flat set of all their
ancestors (excluding the vertices themselves). This function will work with
cycles, since it doesn't compute the dependency order.

```ts
computeAncestorSet(
  originWorkspaces: string[],
  allWorkspaces: Record<string, WorkspaceDependencyObject>,
): Set<string>
```

### Parameters

- `originWorkspaces`: the name of the vertices of which we want to know the
  ancestors
- `allWorkspaces`: a record of workspaces expressed as
  `{ "workspace_name": { workspaceDependencies?: [ 'another_workspace', ...] }, ... }`

### Example

```ts
const workspaces = {
  a: { workspaceDependencies: ['b', 'c'] },
  b: { workspaceDependencies: ['d'] },
  c: { workspaceDependencies: ['b'] },
  d: { workspaceDependencies: undefined },
  e: { workspaceDependencies: ['a', 'b', 'c'] },
};
```

```mermaid
graph TD;
    A-->B;
    A-->C;
    B-->D;
    C-->B;
    E-->A;
    E-->B;
    E-->C;
```

```ts
computeAncestorSet(['d', 'b'], workspaces))
// Result:  new Set(['a', 'c', 'e'])
```

## computeDescendantSet

From one or more vertices in a graph, compute the flat set of all their
dependants (excluding the vertices themselves). This function will work with
cycles, since it doesn't compute the dependency order.

```ts
computeDescendantSet(
  originWorkspaces: string[],
  allWorkspaces: Record<string, WorkspaceDependencyObject>,
): Set<string>
```

### Parameters

- `originWorkspaces`: the name of the vertices of which we want to know the
  ancestors
- `allWorkspaces`: a record of workspaces expressed as
  `{ "workspace_name": { workspaceDependencies?: [ 'another_workspace', ...] }, ... }`

### Example

```ts
const workspaces = {
  a: { workspaceDependencies: ['b', 'c'] },
  b: { workspaceDependencies: ['d'] },
  c: { workspaceDependencies: ['b'] },
  d: { workspaceDependencies: undefined },
  e: { workspaceDependencies: ['a', 'b', 'c'] },
};
```

```mermaid
graph TD;
    A-->B;
    A-->C;
    B-->D;
    C-->B;
    E-->A;
    E-->B;
    E-->C;
```

```ts
computeDescendantSet(['a', 'b'], workspaces))
// Result:  new Set(['c', 'd'])
```

## traverseWorkspaceRelations

Traverses a dependency graph from an array of vertices and returns a map from
vertices to inverse build order. The order can be used to compute a set of build
steps, where dependencies of the same order can be executed in parallel. This
function will throw in case of cycles, with an exception describing the first
cycle found, because the order can't be computed (e.g. if `a` needs `b` to be
built and `b` needs `a` to be built, their respective order will grow to
`Infinity`).

```ts
traverseWorkspaceRelations(
  workspaceNames: string[],
  workspaces: Record<string, WorkspaceDependencyObject>
): OrderedDependencies
```

### Parameters

- `workspaceNames`: an array of start vertices
- `workspaces`: a record of workspaces expressed as
  `{ "workspace_name": { workspaceDependencies?: [ 'another_workspace', ...] }, ... }`

#### Serial order

```ts
workspaces = {
  a: { workspaceDependencies: ['b', 'c'] },
  b: { workspaceDependencies: ['d'] },
  c: { workspaceDependencies: ['b'] },
  d: { workspaceDependencies: undefined },
  e: { workspaceDependencies: ['a', 'b', 'c'] },
};
```

```mermaid
graph TD;
    A-->B;
    A-->C;
    B-->D;
    C-->B;
    E-->A;
    E-->B;
    E-->C;
```

```ts
traverseWorkspaceRelations(['a'], workspaces);
/*
// To build a, first build d, then build d, then build c
Map(
    [
        ['a', 0],
        ['c', 1],
        ['b', 2],
        ['d', 3],
    ]
)
*/
```

#### Serial / parallel order

```ts
workspaces = {
  a: { workspaceDependencies: ['b', 'c'] },
  b: { workspaceDependencies: ['d'] },
  c: { workspaceDependencies: ['d'] },
  d: { workspaceDependencies: undefined },
};
```

```mermaid
graph TD;
    A-->B;
    A-->C;
    B-->D;
    C-->D;
```

```ts
traverseWorkspaceRelations(['a'], workspaces);
/*
// To build a, first build d, then c and b can be built parallely
Map(
    [
        ['a', 0],
        ['c', 1],
        ['b', 1],
        ['d', 2],
    ]
)
*/
```

#### Cycle

```ts
workspaces = {
  a: { workspaceDependencies: ['b'] },
  b: { workspaceDependencies: ['c'] },
  c: { workspaceDependencies: ['b'] },
};
```

```mermaid
graph TD;
    A-->B;
    B-->C;
    C-->B;
```

```ts
traverseWorkspaceRelations(['a'], workspaces);
/* Will throw */
```

### Multiple vertices with some descendants in common

```ts
workspaces: = {
  a: { workspaceDependencies: ['b', 'c'] },
  b: { workspaceDependencies: ['c'] },
  c: { workspaceDependencies: ['d', 'e'] },
  d: { workspaceDependencies: ['e'] },
  e: { workspaceDependencies: [] },
  f: { workspaceDependencies: ['d', 'g'] },
  g: { workspaceDependencies: ['h', 'i'] },
  h: { workspaceDependencies: ['e', 'i'] },
  i: { workspaceDependencies: [] },
};
```

```mermaid
graph TD;
    A-->B;
    A-->C;
    B-->C;
    C-->D;
    C-->E;
    D-->E;
    F-->D;
    F-->G;
    G-->H;
    G-->I;
    H-->E;
    H-->I;
```

```ts
/*
Map([
  ['a', 0],
  ['b', 1],
  ['c', 2],
  ['d', 3],
  ['e', 4],
  ['f', 0],
  ['g', 1],
  ['h', 2],
  ['i', 3],
]),
*/
```

### Multiple vertices with no descendants in common

```ts
workspaces = {
  a: { workspaceDependencies: ['b', 'c'] },
  b: { workspaceDependencies: ['c'] },
  c: { workspaceDependencies: ['d', 'e'] },
  d: { workspaceDependencies: ['e'] },
  e: { workspaceDependencies: [] },
  f: { workspaceDependencies: ['g'] },
  g: { workspaceDependencies: ['h', 'i'] },
  h: { workspaceDependencies: ['i'] },
  i: { workspaceDependencies: [] },
};
```

```mermaid
graph TD;
    A-->B;
    A-->C;
    B-->C;
    C-->D;
    C-->E;
    D-->E;
    F-->G;
    G-->H;
    G-->I;
    H-->I;
```

```ts
traverseWorkspaceRelations(['a', 'f'], workspaces))

/*
Lots of parallelism here, since we can build descendants of a at the same time as descendants of f
Map(
  [
    ['a', 0],
    ['b', 1],
    ['c', 2],
    ['d', 3],
    ['e', 4],
    ['f', 0],
    ['g', 1],
    ['h', 2],
    ['i', 3],
  ]),
);
*/
```

## invertDependencyDirection

Takes a dependency graph and inverts the edge direction. The result is a graph
describing the same information as the original but with the relationship
inverted. Useful to produce an ancestor graph from a descendant graph and vice
versa. Please note that, to optimize for performance, this function can produce
graphs in which some `workspaceDependencies` are `undefined` (instead of empty
arrays).

### Parameters

- `workspaces`: a record of workspaces expressed as
  `{ "workspace_name": { workspaceDependencies?: [ 'another_workspace', ...] }, ... }`

### Example

```ts
const workspaces = {
  a: { workspaceDependencies: ['b', 'c'] },
  b: { workspaceDependencies: ['d'] },
  c: { workspaceDependencies: ['b'] },
  d: { workspaceDependencies: undefined },
  e: { workspaceDependencies: ['a', 'b', 'c'] },
};
```

```mermaid
graph TD;
    A-->B;
    A-->C;
    B-->D;
    C-->B;
    E-->A;
    E-->B;
    E-->C;
```

```ts
invertDependencyDirection(workspaces);
```

#### Resulting (inverted) graph

```ts
{
    a: { workspaceDependencies: ['e'] },
    b: { workspaceDependencies: ['a', 'c', 'e'] },
    c: { workspaceDependencies: ['a', 'e'] },
    d: { workspaceDependencies: ['b'] },
}
```

```mermaid
graph TD;
    A-->E;
    B-->A;
    B-->C;
    B-->E;
    C-->A;
    C-->E;
    D-->B;
```

### CODE REVIEW

File: packages/workspace-resolver/src/resolve-dependencies/README.md [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+
+- `originWorkspaces`: the name of the vertices of which we want to know the
+  ancestors
+- `allWorkspaces`: a record of workspaces expressed as
+  `{ "workspace_name": { workspaceDependencies?: [ 'another_workspace', ...] }, ... }`
+
+### Example
+
+```ts
+const workspaces = {
```
Comment (sgb-io) [inline packages/workspace-resolver/src/resolve-dependencies/README.md:26]: Shouldn't everything that depends on B also depend on D? Not directly, but they're still ancestors, via B

Comment (sgb-io) [reply packages/workspace-resolver/src/resolve-dependencies/README.md:26]: Nvm, misread the docs

File: packages/workspace-resolver/src/resolve-dependencies/README.md [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+```
+
+```ts
+computeDescendantSet(['a', 'b'], workspaces))
+// Result:  new Set(['c', 'd'])
+```
+
+## traverseWorkspaceRelations
+
+Traverses a dependency graph from a single vertix and returns an map from
```
Comment (sgb-io) [inline packages/workspace-resolver/src/resolve-dependencies/README.md:101]: Small typo: `an map` => `a map`

Comment (sgb-io) [reply packages/workspace-resolver/src/resolve-dependencies/README.md:101]: Also should the term be `vertex`?

File: packages/workspace-resolver/src/resolve-dependencies/index.ts [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    workspaceNames,
+    invertDependencyDirection(allWorkspaces),
+  );
+}
+
+// This function takes a tree of dependencies (dependant -> child dependencies)
+// and returns an equivalent tree where the relation's direction is inverted
+// (dependency -> parent dependencies)
+// This allows us to use the same algorithm to query ancestors or descendants.
+export function invertDependencyDirection(
```
Comment (sgb-io) [inline packages/workspace-resolver/src/resolve-dependencies/index.ts:50]: Does `invertDependencyDirection` currently handle cycles? I'm wondering if it's currently assuming acyclic graphs.

I guess if not, it should throw, like how `traverseWorkspaceRelations` does

**Comment by sgb-io** [comment]
Rest LGTM 👍

File: packages/workspace-resolver/src/resolve-dependencies/index.ts (hunk 3 above)
Comment (cristiano-belloni) [reply packages/workspace-resolver/src/resolve-dependencies/index.ts:50]: It handles cycles, as it just inverts the tree representation of a graph into another tree representation.

Comment (sgb-io) [reply packages/workspace-resolver/src/resolve-dependencies/index.ts:50]: Fair enough 👍

File: packages/workspace-resolver/src/resolve-dependencies/README.md (hunk 2 above)
Comment (sgb-io) [inline packages/workspace-resolver/src/resolve-dependencies/README.md:101]:
```suggestion
Traverses a dependency graph from a single vertex and returns a map from
```

File: packages/workspace-resolver/src/resolve-dependencies/README.md [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+```ts
+traverseWorkspaceRelations(
+  workspaceName: string,
+  workspaces: Record<string, WorkspaceDependencyObject>
+): OrderedDependencies
+```
+
+### Parameters
+
+- `workspaceName`: the name of the starting vertix
```
Comment (sgb-io) [inline packages/workspace-resolver/src/resolve-dependencies/README.md:118]:
```suggestion
- `workspaceName`: the name of the starting vertex
```

**Review by sgb-io — APPROVED** [review]

**Pull request merged**